keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,578 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The current Job controller currently relies on completed Pods to not be removed
+in order to track the Job completion status. This proposal presents an
+alternative implementation for Job tracking that does not have this dependency.
+
+## Motivation
+
+The current approach of relying on the Pods existence is problematic for Jobs
+that require a big number of completions or for clusters with too many Jobs
+running at the same time. The finished Pods cannot be removed until the entire
+Job completes, even if the Pods failed.
+
+Furthermore, once the number of finished Pods reaches a threshold, the Pod
+garbage collection controller starts removing Pods. Today, the Job controller
+relies on the garbage collector having a big threshold.
+
+### Goals
+
+- Perform Job completion and failures tracking without relying on lingering
+  Pods.
+
+### Non-Goals
+
+- Remove Pods once they have been accounted for.
+
+## Proposal
+
+The Job controller creates Pods with a finalizer to prevent finished Pods to be
+removed by the garbage collection. The Job controller removes the finalizer from
+the finished Pods once it has accounted for them. In subsequent Job syncs, the
+controller ignores finished Pods that don't have a finalizer.
+
+### Notes/Constraints/Caveats (Optional)
+
+Due to the lack of support of atomic changes across kubernetes objects, an
+intermediate state is necessary. Before removing the Pod finalizers,
+the controller adds finished Pods to a list in the Job status. After removing
+the Pod finalizers, the controller clears the list and updates the counters.
+
+### Risks and Mitigations
+
+1. The new implementation produces new API calls: one per Pod to remove
+   finalizers and one for each Job sync, to track the intermediate state.
+   However, the ability to handle higher number of Pods and Jobs at a given
+   point justifies the added cost of the new API calls.
+2. Changes in the status not produced by the Job controller in
+   kube-controller-manager could affect the Job tracking. Cluster administrators
+   should make sure to protect the Job status endpoint via RBAC.
+
+## Design Details
+
+### API changes
+
+The Job status gets a new struct to hold the uncounted Pods before they are
+added to the counters.
+
+```golang
+type JobStatus struct {
+    Succeeded int32
+    Failed    int32
+    ...
+
+    // UncountedPodUIDs holds UIDs of Pods that have finished but haven't
+    // been accounted in the counters.
+    UncountedPodUIDs UncountedPodUIDs
```

Comment:
Please ensure that the size of that is bounded (controller won't put more than e.g. 100 pods in that list).
If there are more pods to remove - it should be done in few steps.

[The cost of job update api calls is amortized by finalizers removal anyway - so that sounds fine.]